src/ProtoBuf.FSharp/CodeGen.fs

```diff
@@ -0,0 +1,133 @@
+module private ProtoBuf.FSharp.CodeGen
+
+open System
+open FSharp.Reflection
+open System.Collections.Concurrent
+open System.Reflection
+open System.Reflection.Emit
+
+
+let private emitFieldAssignments (gen : ILGenerator, assigns : struct (FieldInfo * MethodInfo)[]) =
+    for (fi, getValue) in assigns do
+        if fi.IsStatic then
+            gen.EmitCall(OpCodes.Call, getValue, null)
+            gen.Emit(OpCodes.Stsfld, fi)
+        else
+            gen.Emit(OpCodes.Dup)
+            gen.EmitCall(OpCodes.Call, getValue, null)
+            gen.Emit(OpCodes.Stfld, fi)
+
+
+let private emitFactory (resultType : Type, assigns) =
+    let factoryMethod = DynamicMethod("factory_" + resultType.FullName, resultType, [| |], true)
+    let gen = factoryMethod.GetILGenerator()
+
+    match resultType.GetConstructor [| |] with
+    | null when FSharpType.IsRecord (resultType, true) ->
+        for pi in FSharpType.GetRecordFields (resultType, true) do
+            let tp = pi.PropertyType
+            match ZeroValues.calculateIfApplicable tp with
+            | Some getValue -> gen.EmitCall(OpCodes.Call, getValue, null)
+            | _ when tp.IsValueType ->
+                let cell = gen.DeclareLocal(tp)
+                gen.Emit(OpCodes.Ldloca_S, cell)
+            | _ -> gen.Emit(OpCodes.Ldnull)
+        let ctr = FSharpValue.PreComputeRecordConstructorInfo (resultType, true)
+        gen.Emit (OpCodes.Newobj, ctr)
+    | null ->
+        gen.Emit(OpCodes.Ldtoken, resultType)
+        gen.EmitCall(OpCodes.Call, MethodHelpers.getMethodInfo <@ Runtime.Serialization.FormatterServices.GetUninitializedObject @> [||], null)
+        emitFieldAssignments (gen, assigns)
+    | ctr ->
+        gen.Emit (OpCodes.Newobj, ctr)
+        emitFieldAssignments (gen, assigns)
+
+    gen.Emit(OpCodes.Ret)
+    factoryMethod :> MethodInfo
+
+
+let private emitRecordSurrogate (surrogateModule : ModuleBuilder, recordType : Type, useValueTypeSurrogate : bool) =
+    let surrogateType =
+        let name = recordType.Name + "Surrogate"
```

Comment:
Small question - where is the namespace for the type determined? The reason I ask is that it is very common to have duplicate types in schemas (e.g. Types.V1 and Types.V2). We don't want clashes on the generated type name; but I would prefer it to be deterministic as well/repeatable to protect against double registration.

Reply:
I thought for some reason that name for types emitted with reflection is irrelevant. Apparently that's not true (and in any case it would've appeared in debugger and stack traces)